### PR TITLE
Create a component for graphs for controllers overview

### PR DESCRIPTION
**Pull request by anthonydillon**
## Done
Create a component for graphs for controllers overview

## QA
- Pull code
- Run `./run clean && ./run serve`
- Open http://0.0.0.0:8036/controllers
- Check that the charts look like [the design](https://usercontent.irccloud-cdn.com/file/DgXu9v8n/mockup.jpg)
- Check it works responsively
- Check the numbers add up

_The chart is missing the central percentage as the data is below and my change_

## Details
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1065

## Screenshots
[if relevant, include a screenshot]
image

### CODE REVIEW

**Comment by hatched** [comment]
Before this lands, lets discuss if the 1.5MB of JS for d3 is absolutely necessary.

**Comment by anthonydillon** [comment]
I can look at getting a lighter version but would like to continue to load d3 from NPM instead of downloading a custom build. This will keep us up to date. Also, won't we need d3 for the bundle Vis anyway?

File: src/app/selectors.js [hunk 1]
```diff
@@ -341,6 +419,33 @@ export const getGroupedModelDataByStatus = createSelector(
...
+  Returns the machine instances sorted by status.
+  @returns {Function} The memoized selector to return the sorted machine instances.
+*/
+export const getGroupedMachinesDataByStatus = createSelector(
+  getModelData,
+  groupMachinesByStatus
+);
+
+/**
+  Returns the machine instances sorted by status.
```
Comment (barrymcgee) [inline src/app/selectors.js]: Update doc string

File: src/pages/Controllers/ControllerChart/_controller-chart.scss [hunk 2]
```diff
@@ -0,0 +1,52 @@
+$breakpoint-medium: 772px;
+
+.p-chart {
+  @media (min-width: $breakpoint-medium) {
+    display: flex;
+    align-items: center;
+
+    .p-chart__chart {
```
Comment (barrymcgee) [inline src/pages/Controllers/ControllerChart/_controller-chart.scss]: Use parent selector here?

File: src/pages/Controllers/ControllerOverview/_controller-overview.scss [hunk 3]
```diff
@@ -0,0 +1,8 @@
+.controllers-overview {
+  overflow-x: auto;
+  width: 100%;
+
+  .controller-overview__container {
```
Comment (barrymcgee) [inline src/pages/Controllers/ControllerOverview/_controller-overview.scss]:
```suggestion
  .&__container {
```

File: package.json [hunk 4]
```diff
@@ -54,6 +54,7 @@
     "@craco/craco": "^5.6.2",
     "async-limiter": "^2.0.0",
     "classnames": "^2.2.6",
+    "d3": "^5.15.0",
```
Comment (barrymcgee) [inline package.json]:
```suggestion
    "d3": "5.15.0",
```

File: src/app/selectors.js [hunk 5]
```diff
@@ -169,6 +172,81 @@ const groupModelsByStatus = modelData => {
   return grouped;
 };
 
+/**
+  Returns a grouped collection of machine instances.
+  @param {Object} modelData
+  @returns {Function} The grouped machine instances.
```
Comment (barrymcgee) [inline src/app/selectors.js]:
```suggestion
  @returns {Object} The grouped machine instances.
```

File: src/app/selectors.js [hunk 6]
```diff
@@ -169,6 +172,81 @@ const groupModelsByStatus = modelData => {
...
+  for (let modelUUID in modelData) {
+    const model = modelData[modelUUID];
+    for (let applicationID in model.applications) {
+      const application = model.applications[applicationID];
+      grouped[getApplicationStatusGroup(application).status].push(application);
+    }
+  }
+  return grouped;
+};
+
```
Comment (barrymcgee) [inline src/app/selectors.js]: These three functions are so similar that I feel you should create a higher-order function for them where you could just pass `modelData` and the name of the collection?

File: src/components/DonutChart/DonutChart.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      .attr("height", height)
+      .append("g")
+      .attr("transform", "translate(" + width / 2 + "," + height / 2 + ")");
+
+    const data = chartData;
+
+    const color = d3
+      .scaleOrdinal()
+      .domain(data)
+      .range(["#c7162b", "#f99b11", "#cdcdcd"]);
```
Comment (barrymcgee) [inline src/components/DonutChart/DonutChart.js]: Just an idea but if we assign these colors to CSS variables, we can then access them via JS

https://davidwalsh.name/css-variables-javascript

File: src/components/DonutChart/DonutChart.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    const pie = d3.pie().value(function(d) {
+      return d.value;
+    });
+    const data_ready = pie(d3.entries(data));
+
+    svg
+      .selectAll("whatever")
```
Comment (barrymcgee) [inline src/components/DonutChart/DonutChart.js]: This appears unused anywhere?

File: src/components/DonutChart/DonutChart.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    const color = d3
+      .scaleOrdinal()
+      .domain(data)
+      .range(["#c7162b", "#f99b11", "#cdcdcd"]);
+
+    const pie = d3.pie().value(function(d) {
+      return d.value;
+    });
+    const data_ready = pie(d3.entries(data));
```
Comment (barrymcgee) [inline src/components/DonutChart/DonutChart.js]: Variables are usually camelCase

File: src/pages/Controllers/ControllerChart/_controller-chart.scss [hunk 10]
```diff
@@ -0,0 +1,52 @@
+$breakpoint-medium: 772px;
```
Comment (barrymcgee) [inline src/pages/Controllers/ControllerChart/_controller-chart.scss]: Override by 4px seems like a code smell.

File: src/pages/Controllers/ControllerChart/_controller-chart.scss [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    &.is-blocked::before {
+      background-color: #c7162b;
+    }
+
+    &.is-alert::before {
+      background-color: #f99b11;
+    }
+
+    &.is-running::before {
+      background-color: #cdcdcd;
```
Comment (barrymcgee) [inline src/pages/Controllers/ControllerChart/_controller-chart.scss]: Use Vanilla vars?

**Comment by spencerbygraves** [comment]
Thanks @anthonydillon. You can refer to the [design on Zeplin](https://zpl.io/2jR9Yrp) for the correct graph size and font spacing.

The graphs don't seem to have any dithering, so the edges are rough. Can that be changed?

Thanks.

File: src/pages/Controllers/ControllerChart/_controller-chart.scss (hunk 2 above)
Comment (anthonydillon) [reply src/pages/Controllers/ControllerChart/_controller-chart.scss]: Done

File: src/app/selectors.js (hunk 5 above)
Comment (anthonydillon) [reply src/app/selectors.js]: Done

File: package.json (hunk 4 above)
Comment (anthonydillon) [reply package.json]: Done

File: src/components/DonutChart/DonutChart.js (hunk 9 above)
Comment (anthonydillon) [reply src/components/DonutChart/DonutChart.js]: Done

File: src/pages/Controllers/ControllerChart/_controller-chart.scss [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    &::before {
+      position: absolute;
+      left: 0;
+      top: 0.65rem;
+      content: "";
+      display: inline-block;
+      width: 0.5rem;
+      height: 0.5rem;
+      border-radius: 50%;
```
[the remark this answers is not in the document]
Comment (anthonydillon) [reply src/pages/Controllers/ControllerChart/_controller-chart.scss]: Done

File: src/pages/Controllers/ControllerChart/_controller-chart.scss (hunk 11 above)
Comment (anthonydillon) [reply src/pages/Controllers/ControllerChart/_controller-chart.scss]: Done, thanks for the tip.

**Comment by anthonydillon** [comment]
@hatched I would like to go ahead with this and I will create [an issue to investigate](https://github.com/canonical-web-and-design/jaas-dashboard/issues/284) a lighter weight bundle in the future. The time to load the API data far out weights the time taken to load an extra MB. That said I am keen to reduce the size and add lazy loading of charting dependencies which can all be added in follow up performance work.

File: src/pages/Controllers/ControllerChart/ControllerChart.test.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import { MemoryRouter } from "react-router";
+import { Provider } from "react-redux";
+import configureStore from "redux-mock-store";
+
+import ControllerChart from "./ControllerChart";
+
+const mockStore = configureStore([]);
+
+describe("Controllers chart", () => {
+  it("renders if totals correctly", () => {
```
Comment (hatched) [inline src/pages/Controllers/ControllerChart/ControllerChart.test.js]: remove "if"

File: src/pages/Controllers/ControllerChart/ControllerChart.test.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    const store = mockStore({
```
Comment (hatched) [inline src/pages/Controllers/ControllerChart/ControllerChart.test.js]: These `store`s aren't required nor is the `MemoryRouter` or `Provider` for any of the tests for this component.

File: src/pages/Controllers/ControllerChart/ControllerChart.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import React from "react";
+
+import DonutChart from "components/DonutChart/DonutChart";
+
+import { pluralize } from "app/utils";
+
+import "./_controller-chart.scss";
+
+export default function ControllerChart({ chartData, totalCount, totalLabel }) {
```
Comment (hatched) [inline src/pages/Controllers/ControllerChart/ControllerChart.js]: the `totalCount` should probably be derived from the data.

File: src/pages/Controllers/ControllerOverview/ControllerOverview.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+export default function ControllersOverview() {
+  const groupedMachinesDataByStatus = useSelector(
+    getGroupedMachinesDataByStatus
+  );
+  const groupedApplicationsDataByStatus = useSelector(
+    getGroupedApplicationsDataByStatus
+  );
+  const groupedUnitsDataByStatus = useSelector(getGroupedUnitsDataByStatus);
+
+  const { machinesCount, applicationCount, unitCount } = useSelector(
```
Comment (hatched) [inline src/pages/Controllers/ControllerOverview/ControllerOverview.js]: This component is running a lot of loops over the same data extracting different data. It'll be more efficient to count the lengths you are already accessing below.

Also if you're going to derive the counts from the data in the `ControllerChat` component then these values aren't necessary.

**Comment by spencerbygraves** [comment]
Spoke to @anthonydillon yesterday, as the design is probably going to change a bit more, I'm happy to add +1 to this and come back to it. Thanks.

File: src/app/utils.js [hunk 17]
```diff
@@ -89,16 +89,52 @@ export const getApplicationStatusGroup = application => {
...
-  @returns {Object} The status of the unit and any relevent messaging.
+  Returns the status for the machine.
+  @param {Object} machine The machine to check the status of in the
+    format stored in the redux store.
+  @returns {Object} The status of the machine and any relevent messaging.
 */
-export const getUnitStatusGroup = unit => {
+export const getMachineStatusGroup = machine => {
+  // Possible "blocked" or error states in machine statuses.
+  const blocked = ["down"];
```
[the remark this answers is not in the document]
Comment (hatched) [reply src/app/utils.js]: As this naming is used everywhere I'll leave this for now and we can revisit later.

**Pull request merged**